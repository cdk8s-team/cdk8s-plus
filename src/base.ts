import { ApiObjectMetadata, ApiObject, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiResource, IApiEndpoint } from './api-resource.generated';
import * as r from './role';
import * as rb from './role-binding';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   */
  readonly metadata?: ApiObjectMetadata;
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  readonly apiVersion: string;

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io")
   */
  readonly apiGroup: string;

  /**
   * The object kind (e.g. "Deployment").
   */
  readonly kind: string;
}

/**
 * Base class for all Kubernetes objects in stdk8s. Represents a single
 * resource.
 */
export abstract class Resource extends Construct implements IResource, IApiResource, IApiEndpoint {

  /**
   * The underlying cdk8s API object.
   */
  protected abstract readonly apiObject: ApiObject;

  public get metadata(): ApiObjectMetadataDefinition {
    return this.apiObject.metadata;
  }

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  public get apiVersion(): string {
    return this.apiObject.apiVersion;
  }

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io").
   */
  public get apiGroup(): string {
    return this.apiObject.apiGroup;
  }

  /**
   * The object kind (e.g. "Deployment").
   */
  public get kind(): string {
    return this.apiObject.kind;
  }

  public get resourceName(): string | undefined {
    return this.name;
  }

  public readonly abstract resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }

  /**
   * Grants the list of subjects permissions to read this resource.
   */
  public grantRead(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowRead(this);
    return role.bind(...subjects);
  }

  /**
   * Grants the list of subjects permissions to read and write this resource.
   */
  public grantReadWrite(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowReadWrite(this);
    return role.bind(...subjects);
  }

}