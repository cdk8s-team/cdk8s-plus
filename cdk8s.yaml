output: dist
pluginsDirectory: /home/runner/.cdk8s/plugins
imports:
  - k8s@1.27.0
  - k8s@1.28.0
