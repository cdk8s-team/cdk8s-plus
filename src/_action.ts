import { Container } from './container';
import * as k8s from './imports/k8s';

/**
 * Utility class to implement the conversion between our API and the k8s action
 * structure. Used both for probes and handlers.
 *
 * @internal
 */
export class Action {

  public static fromTcpSocket(container: Container, options: { port?: number; host?: string } = {}): k8s.TcpSocketAction {
    return {
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
      host: options.host,
    };
  }

  public static fromCommand(command: string[]): k8s.ExecAction {
    return { command };
  }

  public static fromHttpGet(container: Container, path: string, options: { port?: number } = { }): k8s.HttpGetAction {
    return {
      path,
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
    };
  }
}