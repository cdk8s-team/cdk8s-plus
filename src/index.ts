export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './job';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume';
export * from './ingress';
export * from './probe';
export * from './pvc';
export * from './pv';
export * from './handler';
export * from './workload';
export * from './daemon-set';
export * from './role';
export * from './role-binding';
export * from './namespace';

export * from './api-resource.generated';
