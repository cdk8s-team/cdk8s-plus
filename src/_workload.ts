import { ApiObjectMetadata, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import { Container, ContainerProps } from './container';
import { HostAlias, PodSecurityContext, PodSpec, PodSpecProps, RestartPolicy } from './pod';
import { IServiceAccount } from './service-account';
import { Volume } from './volume';

/**
 * Properties for `Workload`.
 */
export interface WorkloadProps extends ResourceProps, PodSpecProps {

  /**
   * The pod metadata of this workload.
   */
  readonly podMetadata?: ApiObjectMetadata;

}

export abstract class Workload extends Resource {

  protected readonly podSpec: PodSpec;

  /**
   * The metadata of pods in this deployment.
   */
  public readonly podMetadata: ApiObjectMetadataDefinition;

  constructor(scope: Construct, id: string, props: WorkloadProps = {}) {
    super(scope, id);

    this.podSpec = new PodSpec(props);
    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
  }

  public get containers(): Container[] {
    return this.podSpec.containers;
  }

  public get initContainers(): Container[] {
    return this.podSpec.initContainers;
  }

  public get volumes(): Volume[] {
    return this.podSpec.volumes;
  }

  public get restartPolicy(): RestartPolicy | undefined {
    return this.podSpec.restartPolicy;
  }

  public get serviceAccount(): IServiceAccount | undefined {
    return this.podSpec.serviceAccount;
  }

  public get securityContext(): PodSecurityContext {
    return this.podSpec.securityContext;
  }

  public get hostAliases(): HostAlias[] {
    return this.podSpec.hostAliases;
  }

  public addContainer(container: ContainerProps): Container {
    return this.podSpec.addContainer(container);
  }

  public addInitContainer(container: ContainerProps): Container {
    return this.podSpec.addInitContainer(container);
  }

  public addVolume(volume: Volume): void {
    return this.podSpec.addVolume(volume);
  }

  public addHostAlias(hostAlias: HostAlias): void {
    return this.podSpec.addHostAlias(hostAlias);
  }

}