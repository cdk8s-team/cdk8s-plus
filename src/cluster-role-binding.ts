import { ApiObject, Lazy } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import * as clusterrole from './cluster-role';
import * as k8s from './imports/k8s';
import * as rb from './role-binding';
import { filterUndefined } from './utils';

/**
 * Properties for `ClusterRoleBinding`.
 */
export interface ClusterRoleBindingProps extends ResourceProps {
  /**
   * The role to bind to.
   */
  readonly role: clusterrole.IClusterRole;
}

/**
 * A ClusterRoleBinding grants permissions cluster-wide to a user or
 * set of users.
 */
export class ClusterRoleBinding extends Resource {
  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  public readonly resourceType = 'clusterrolebindings';

  public readonly role: clusterrole.IClusterRole;

  private readonly _subjects: Array<rb.ISubject>;

  constructor(scope: Construct, id: string, props: ClusterRoleBindingProps) {
    super(scope, id);

    this.role = props.role;

    this._subjects = new Array<rb.ISubject>();

    this.apiObject = new k8s.KubeClusterRoleBinding(this, 'Resource', {
      metadata: props.metadata,
      subjects: Lazy.any({ produce: () => this.synthesizeSubjects() }),
      roleRef: Lazy.any({ produce: () => this.synthesizeRoleRef() }),
    });
  }

  /**
   * Adds a subject to the role.
   * @param subjects The subjects to add
   */
  public addSubjects(...subjects: rb.ISubject[]): void {
    for (const subject of subjects) {
      this._subjects.push(subject);
    }
  }

  public get subjects() {
    return [...this._subjects];
  }

  private synthesizeRoleRef(): k8s.RoleRef {
    return {
      apiGroup: this.role.apiGroup,
      kind: this.role.kind,
      name: this.role.name,
    };
  }

  private synthesizeSubjects(): k8s.Subject[] {
    return this._subjects.map((subject) => filterUndefined({
      apiGroup: subject.apiGroup === 'core' ? '' : subject.apiGroup,
      kind: subject.kind,
      name: subject.name,
      namespace: subject.namespace,
    }));
  }
}
