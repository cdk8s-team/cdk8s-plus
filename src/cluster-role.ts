import { ApiObject, Lazy, Names } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiEndpoint } from './api-resource.generated';
import * as base from './base';
import * as clusterrb from './cluster-role-binding';
import * as k8s from './imports/k8s';
import * as role from './role';
import * as rb from './role-binding';

/**
 * Represents a cluster-level role.
 */
export interface IClusterRole extends base.IResource {}

/**
 * Policy rule of a `ClusterRole.
 */
export interface ClusterRolePolicyRule {

  /**
   * Verbs to allow. (e.g ['get', 'watch'])
   */
  readonly verbs: string[];

  /**
   * Endpoints this rule applies to. Can be either api resources
   * or non api resources.
   */
  readonly endpoints: IApiEndpoint[];
}

/**
 * Properties for `ClusterRole`.
 */
export interface ClusterRoleProps extends base.ResourceProps {

  /**
   * A list of rules the role should allow.
   *
   * @default []
   */
  readonly rules?: ClusterRolePolicyRule[];

  /**
   * Specify labels that should be used to locate ClusterRoles, whose rules
   * will be automatically filled into this ClusterRole's rules.
   */
  readonly aggregationLabels?: { [key: string]: string };
}

/**
 * ClusterRole is a cluster level, logical grouping of PolicyRules that can be
 * referenced as a unit by a RoleBinding or ClusterRoleBinding.
 */
export class ClusterRole extends base.Resource implements IClusterRole, role.IRole {

  /**
   * Imports a role from the cluster as a reference.
   * @param name The name of the role resource.
   */
  public static fromClusterRoleName(name: string): IClusterRole {
    return {
      apiGroup: 'rbac.authorization.k8s.io',
      name,
      ...k8s.KubeClusterRole.GVK,
    };
  }

  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  public readonly resourceType = 'clusterroles';

  private readonly _labelSelector: Record<string, string> = {};
  private readonly _rules: Array<k8s.PolicyRule> = [];

  constructor(scope: Construct, id: string, props: ClusterRoleProps = {}) {
    super(scope, id);

    this.apiObject = new k8s.KubeClusterRole(this, 'Resource', {
      metadata: props.metadata,
      rules: Lazy.any({ produce: () => this._rules }),
      aggregationRule: Lazy.any({ produce: () => this.synthesizeAggregationRules() }),
    });

    for (const rule of props.rules ?? []) {
      this.allow(rule.verbs, ...rule.endpoints);
    }

    for (const [key, value] of Object.entries(props.aggregationLabels ?? {})) {
      this.aggregate(key, value);
    }
  }

  /**
   * Add permission to perform a list of HTTP verbs on a collection of
   * resources.
   *
   * @param endpoints The endpoints(s) to apply to
   * @see https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-the-request-verb
   */
  public allow(verbs: string[], ...endpoints: IApiEndpoint[]): void {
    for (const endpoint of endpoints) {

      const resource = endpoint.asApiResource();
      const nonResource = endpoint.asNonApiResource();

      if (resource && nonResource) {
        throw new Error('Endpoint must be either resource or non resource. not both.');
      }

      if (!resource && !nonResource) {
        throw new Error('Endpoint must be either resource or non resource. not neither.');
      }

      if (resource) {
        this._rules.push({
          apiGroups: [resource.apiGroup === 'core' ? '' : resource.apiGroup],
          resources: [resource.resourceType],
          resourceNames: resource.resourceName ? [resource.resourceName] : [],
          verbs,
        });
      }
      if (nonResource) {
        this._rules.push({ verbs, nonResourceUrLs: [nonResource] });
      }
    }
  }

  /**
   * Add "create" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowCreate(...endpoints: IApiEndpoint[]): void {
    this.allow(['create'], ...endpoints);
  }

  /**
   * Add "get" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowGet(...endpoints: IApiEndpoint[]): void {
    this.allow(['get'], ...endpoints);
  }

  /**
   * Add "list" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowList(...endpoints: IApiEndpoint[]): void {
    this.allow(['list'], ...endpoints);
  }

  /**
   * Add "watch" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowWatch(...endpoints: IApiEndpoint[]): void {
    this.allow(['watch'], ...endpoints);
  }

  /**
   * Add "update" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowUpdate(...endpoints: IApiEndpoint[]): void {
    this.allow(['update'], ...endpoints);
  }

  /**
   * Add "patch" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowPatch(...endpoints: IApiEndpoint[]): void {
    this.allow(['patch'], ...endpoints);
  }

  /**
   * Add "delete" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowDelete(...endpoints: IApiEndpoint[]): void {
    this.allow(['delete'], ...endpoints);
  }

  /**
   * Add "deletecollection" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowDeleteCollection(...endpoints: IApiEndpoint[]): void {
    this.allow(['deletecollection'], ...endpoints);
  }

  /**
   * Add "get", "list", and "watch" permissions for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowRead(...endpoints: IApiEndpoint[]): void {
    this.allow(['get', 'list', 'watch'], ...endpoints);
  }

  /**
   * Add "get", "list", "watch", "create", "update", "patch", "delete", and
   * "deletecollection" permissions for the resources.
   *
   * @param endpoints The resource(s) to apply to
   */
  public allowReadWrite(...endpoints: IApiEndpoint[]): void {
    this.allow(['get', 'list', 'watch', 'create', 'update', 'patch', 'delete', 'deletecollection'], ...endpoints);
  }

  /**
   * Aggregate rules from roles matching this label selector.
   */
  public aggregate(key: string, value: string): void {
    this._labelSelector[key] = value;
  }

  /**
   * Combines the rules of the argument ClusterRole into this ClusterRole
   * using aggregation labels.
   * @param rol
   */
  public combine(rol: ClusterRole): void {
    const key = `cdk8s.cluster-role/aggregate-to-${Names.toLabelValue(this)}`;
    const value = 'true';
    rol.metadata.addLabel(key, value);
    this.aggregate(key, value);
  }

  /**
   * Create a RoleBinding that binds the permissions in this ClusterRole
   * to a list of subjects, that will only apply to the given namespace.
   * @param namespace the namespace to limit permissions to.
   * @param subjects a list of subjects to bind to
   */
  public bindInNamespace(namespace: string, ...subjects: rb.ISubject[]): rb.RoleBinding {
    const binding = new rb.RoleBinding(this, `RoleBinding-${namespace}`, {
      metadata: {
        namespace,
      },
      role: this,
    });
    binding.addSubjects(...subjects);
    return binding;
  }

  /**
   * Create a ClusterRoleBinding that binds the permissions in this
   * ClusterRole to a list of subjects, without namespace restrictions.
   * @param subjects a list of subjects to bind to
   */
  public bind(...subjects: rb.ISubject[]): clusterrb.ClusterRoleBinding {
    const binding = new clusterrb.ClusterRoleBinding(this, 'ClusterRoleBinding', {
      role: this,
    });
    binding.addSubjects(...subjects);
    return binding;
  }

  private synthesizeAggregationRules(): k8s.AggregationRule | undefined {
    if (Object.keys(this._labelSelector).length === 0) {
      return undefined;
    }

    return { clusterRoleSelectors: [{ matchLabels: this._labelSelector }] };
  }
}