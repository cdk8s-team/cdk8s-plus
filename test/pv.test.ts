import * as cdk8s from 'cdk8s';
import { Testing } from 'cdk8s';
import * as kplus from '../src';

describe('PersistentVolume', () => {

  test('can grant permissions on imported', () => {

    const chart = Testing.chart();
    const pv = kplus.PersistentVolume.fromPersistentVolumeName(chart, 'Vol', 'vol');

    const role = new kplus.Role(chart, 'Role');
    role.allowRead(pv);

    expect(Testing.synth(chart)).toMatchSnapshot();

  });

  test('defaults', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    expect(vol.accessModes).toBeUndefined();
    expect(vol.reclaimPolicy).toEqual(kplus.PersistentVolumeReclaimPolicy.RETAIN);
    expect(vol.storage).toBeUndefined();
    expect(vol.storageClassName).toBeUndefined();
    expect(vol.mode).toEqual(kplus.PersistentVolumeMode.FILE_SYSTEM);
    expect(vol.mountOptions).toBeUndefined();

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('custom', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      accessModes: [kplus.PersistentVolumeAccessMode.READ_ONLY_MANY, kplus.PersistentVolumeAccessMode.READ_WRITE_MANY],
      mountOptions: ['opt1'],
      reclaimPolicy: kplus.PersistentVolumeReclaimPolicy.DELETE,
      volumeMode: kplus.PersistentVolumeMode.BLOCK,
      storageClassName: 'storage-class',
      storage: cdk8s.Size.gibibytes(50),
      volumeId: 'vol1',
    });

    // base props
    expect(vol.accessModes).toEqual([kplus.PersistentVolumeAccessMode.READ_ONLY_MANY, kplus.PersistentVolumeAccessMode.READ_WRITE_MANY]);
    expect(vol.reclaimPolicy).toEqual(kplus.PersistentVolumeReclaimPolicy.DELETE);
    expect(vol.storage).toEqual(cdk8s.Size.gibibytes(50));
    expect(vol.storageClassName).toEqual('storage-class');
    expect(vol.mode).toEqual(kplus.PersistentVolumeMode.BLOCK);
    expect(vol.mountOptions).toEqual(['opt1']);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('can be imported', () => {

    const chart = Testing.chart();
    const pv = kplus.PersistentVolume.fromPersistentVolumeName(chart, 'Vol', 'vol');
    expect(pv.name).toEqual('vol');

  });

  test('can be reserved with default storage class', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    const claim = vol.reserve();

    expect(claim.accessModes).toBeUndefined();
    expect(claim.storage).toBeUndefined();
    expect(claim.storageClassName).toEqual(vol.storageClassName);

    // validate bi-directional binding
    expect(claim.volume!.name).toEqual(vol.name);
    expect(vol.claim!.name).toEqual(claim.name);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('can be reserved with a custom storage class', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
      storageClassName: 'storage-class',
    });

    const claim = vol.reserve();

    expect(claim.accessModes).toBeUndefined();
    expect(claim.storage).toBeUndefined();
    expect(claim.storageClassName).toEqual(vol.storageClassName);

    // validate bi-directional binding
    expect(claim.volume!.name).toEqual(vol.name);
    expect(vol.claim!.name).toEqual(claim.name);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('reserved claim is created in the same namespace as the volume', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      metadata: { namespace: 'non-default' },
      volumeId: 'vol1',
    });

    const claim = vol.reserve();
    expect(claim.metadata.namespace).toEqual(vol.metadata.namespace);

  });

  test('throws if reserved twice', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    vol.reserve();

    expect(() => vol.reserve()).toThrowError(/There is already a Construct with name 'test-volume-c8db061ePVC'/);

  });

  test('can be bound to a claim at instantiation', () => {

    const chart = cdk8s.Testing.chart();
    const pvc = kplus.PersistentVolumeClaim.fromClaimName(chart, 'Claim', 'claim');
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
      claim: pvc,
    });

    expect(vol.claim!.name).toEqual(pvc.name);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('can be bound to a claim post instantiation', () => {

    const chart = cdk8s.Testing.chart();
    const pvc = kplus.PersistentVolumeClaim.fromClaimName(chart, 'Claim', 'claim');
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    vol.bind(pvc);

    expect(vol.claim!.name).toEqual(pvc.name);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('no-ops if bounded twice to the same claim', () => {

    const chart = cdk8s.Testing.chart();
    const pvc = kplus.PersistentVolumeClaim.fromClaimName(chart, 'Claim', 'claim');
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    vol.bind(pvc);
    vol.bind(pvc);

    expect(vol.claim!.name).toEqual(pvc.name);

  });

  test('throws if bounded twice to different claims', () => {

    const chart = cdk8s.Testing.chart();
    const pvc1 = kplus.PersistentVolumeClaim.fromClaimName(chart, 'Claim1', 'claim1');
    const pvc2 = kplus.PersistentVolumeClaim.fromClaimName(chart, 'Claim2', 'claim2');
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    vol.bind(pvc1);

    expect(() => vol.bind(pvc2)).toThrowError('Cannot bind volume \'test-volume-c8db061e\' to claim \'claim2\' since it is already bound to claim \'claim1\'');

  });

});

describe('AwsElasticBlockStorePersistentVolume', () => {

  test('defaults', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
    });

    expect(vol.volumeId).toEqual('vol1');
    expect(vol.fsType).toEqual('ext4');
    expect(vol.partition).toBeUndefined();
    expect(vol.readOnly).toBeFalsy();

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('custom', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'Volume', {
      volumeId: 'vol1',
      partition: 1,
      readOnly: true,
      fsType: 'ntfs',
    });

    expect(vol.volumeId).toEqual('vol1');
    expect(vol.fsType).toEqual('ntfs');
    expect(vol.partition).toEqual(1);
    expect(vol.readOnly).toBeTruthy();

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

});

describe('AzureDiskPersistentVolume', () => {

  test('defaults', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AzureDiskPersistentVolume(chart, 'Volume', {
      diskName: 'name',
      diskUri: 'uri',
    });

    expect(vol.diskName).toEqual('name');
    expect(vol.diskUri).toEqual('uri');
    expect(vol.cachingMode).toEqual(kplus.AzureDiskPersistentVolumeCachingMode.NONE);
    expect(vol.readOnly).toBeFalsy();
    expect(vol.fsType).toEqual('ext4');
    expect(vol.azureKind).toEqual(kplus.AzureDiskPersistentVolumeKind.SHARED);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('custom', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.AzureDiskPersistentVolume(chart, 'Volume', {
      diskName: 'name',
      diskUri: 'uri',
      cachingMode: kplus.AzureDiskPersistentVolumeCachingMode.READ_ONLY,
      readOnly: true,
      fsType: 'ntfs',
      kind: kplus.AzureDiskPersistentVolumeKind.DEDICATED,
    });

    expect(vol.diskName).toEqual('name');
    expect(vol.diskUri).toEqual('uri');
    expect(vol.cachingMode).toEqual(kplus.AzureDiskPersistentVolumeCachingMode.READ_ONLY);
    expect(vol.readOnly).toBeTruthy();
    expect(vol.fsType).toEqual('ntfs');
    expect(vol.azureKind).toEqual(kplus.AzureDiskPersistentVolumeKind.DEDICATED);

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

});

describe('GCEPersistentDiskPersistentVolume', () => {

  test('defaults', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.GCEPersistentDiskPersistentVolume(chart, 'Volume', {
      pdName: 'name',
    });

    expect(vol.pdName).toEqual('name');
    expect(vol.partition).toBeUndefined();
    expect(vol.readOnly).toBeFalsy();
    expect(vol.fsType).toEqual('ext4');

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

  test('custom', () => {

    const chart = cdk8s.Testing.chart();
    const vol = new kplus.GCEPersistentDiskPersistentVolume(chart, 'Volume', {
      pdName: 'name',
      partition: 1,
      readOnly: true,
      fsType: 'ntfs',
    });

    expect(vol.pdName).toEqual('name');
    expect(vol.partition).toEqual(1);
    expect(vol.readOnly).toBeTruthy();
    expect(vol.fsType).toEqual('ntfs');

    const resources = cdk8s.Testing.synth(chart);
    expect(resources).toMatchSnapshot();

  });

});
