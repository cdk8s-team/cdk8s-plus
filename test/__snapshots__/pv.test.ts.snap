// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AwsElasticBlockStorePersistentVolume custom 1`] = `
Object {
  "awsElasticBlockStore": Object {
    "fsType": "ntfs",
    "partition": 1,
    "readOnly": true,
    "volumeID": "vol1",
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`AwsElasticBlockStorePersistentVolume defaults 1`] = `
Object {
  "awsElasticBlockStore": Object {
    "fsType": "ext4",
    "readOnly": false,
    "volumeID": "vol1",
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`AzureDiskPersistentVolume custom 1`] = `
Object {
  "azureDisk": Object {
    "cachingMode": "ReadOnly",
    "diskName": "name",
    "diskURI": "uri",
    "fsType": "ntfs",
    "kind": "Dedicated",
    "readOnly": true,
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`AzureDiskPersistentVolume defaults 1`] = `
Object {
  "azureDisk": Object {
    "cachingMode": "None",
    "diskName": "name",
    "diskURI": "uri",
    "fsType": "ext4",
    "kind": "Shared",
    "readOnly": false,
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`GCEPersistentDiskPersistentVolume custom 1`] = `
Object {
  "gcePersistentDisk": Object {
    "fsType": "ntfs",
    "partition": 1,
    "pdName": "name",
    "readOnly": true,
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`GCEPersistentDiskPersistentVolume defaults 1`] = `
Object {
  "gcePersistentDisk": Object {
    "fsType": "ext4",
    "pdName": "name",
    "readOnly": false,
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`PersistentVolume can be bound to a claim at instantiation 1`] = `
Object {
  "awsElasticBlockStore": Object {
    "fsType": "ext4",
    "readOnly": false,
    "volumeID": "vol1",
  },
  "claimRef": Object {
    "name": "claim",
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`PersistentVolume can be bound to a claim post instantiation 1`] = `
Object {
  "awsElasticBlockStore": Object {
    "fsType": "ext4",
    "readOnly": false,
    "volumeID": "vol1",
  },
  "claimRef": Object {
    "name": "claim",
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;

exports[`PersistentVolume can be reserved with a custom storage class 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolume",
    "metadata": Object {
      "name": "test-volume-c8db061e",
    },
    "spec": Object {
      "awsElasticBlockStore": Object {
        "fsType": "ext4",
        "readOnly": false,
        "volumeID": "vol1",
      },
      "claimRef": Object {
        "name": "pvc-test-volume-c8db061e",
      },
      "persistentVolumeReclaimPolicy": "Retain",
      "storageClassName": "storage-class",
      "volumeMode": "Filesystem",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": Object {
      "name": "pvc-test-volume-c8db061e",
    },
    "spec": Object {
      "storageClassName": "storage-class",
      "volumeMode": "Filesystem",
      "volumeName": "test-volume-c8db061e",
    },
  },
]
`;

exports[`PersistentVolume can be reserved with default storage class 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolume",
    "metadata": Object {
      "name": "test-volume-c8db061e",
    },
    "spec": Object {
      "awsElasticBlockStore": Object {
        "fsType": "ext4",
        "readOnly": false,
        "volumeID": "vol1",
      },
      "claimRef": Object {
        "name": "pvc-test-volume-c8db061e",
      },
      "persistentVolumeReclaimPolicy": "Retain",
      "volumeMode": "Filesystem",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": Object {
      "name": "pvc-test-volume-c8db061e",
    },
    "spec": Object {
      "volumeMode": "Filesystem",
      "volumeName": "test-volume-c8db061e",
    },
  },
]
`;

exports[`PersistentVolume custom 1`] = `
Object {
  "accessModes": Array [
    "ReadOnlyMany",
    "ReadWriteMany",
  ],
  "awsElasticBlockStore": Object {
    "fsType": "ext4",
    "readOnly": false,
    "volumeID": "vol1",
  },
  "capacity": Object {
    "storage": "50Gi",
  },
  "mountOptions": Array [
    "opt1",
  ],
  "persistentVolumeReclaimPolicy": "Delete",
  "storageClassName": "storage-class",
  "volumeMode": "Block",
}
`;

exports[`PersistentVolume defaults 1`] = `
Object {
  "awsElasticBlockStore": Object {
    "fsType": "ext4",
    "readOnly": false,
    "volumeID": "vol1",
  },
  "persistentVolumeReclaimPolicy": "Retain",
  "volumeMode": "Filesystem",
}
`;
